### ISSUE TITLE
calculate the submit threshold by priority

### ISSUE THREAD

**Issue by ticoann** [dmwm/WMCore#7123]
Same as freeslots() function, so higher priority jobs will be submitted always even if pending slots are filled with lower priority.  

This still need a discussion with Alan @amaltaro

**Comment by amaltaro** [comment]
Honestly, I'm not sure we really want to add this micro-management to the JobSubmitter code as well.
The WorkQueue code is a classic example that has so many tweaks that makes it very hard to understand and maintain...

**Comment by hufnagel** [comment]
The more complex an implementation is, the more difficult it is to get the behavior out of it that you want without introducing undesired edge cases. You keep adding more tweaks to fix these cases that introduce more edge cases etc etc.

I would suggest to at least think about some modification/simplification that gets us what we want first.

**Comment by amaltaro** [comment]
I'm running a scale test right now against highprio team.
 1. lowprio workflow sent everywhere
 2. highprio workflow sent to PIC and MIT only.

Then looking at the resource control of the agent (cutting only what matters):
```
T2_US_MIT - 586 running, 633 pending, 6399 running slots total, 1279 pending slots total, Site is Normal:
  Production - 586 running, 633 pending, 6399 max running, 1151 max pending, priority 0
    32 - /amaltaro_MonteCarlo_eff_ScaleTest_TEST_160829_100904_2175/Production
    1187 - /amaltaro_MonteCarlo_HighPrio_ScaleTest_TEST_160829_133434_146/Production
and
T1_ES_PIC - 738 running, 265 pending, 1692 running slots total, 338 pending slots total, Site is Normal:
  Production - 737 running, 263 pending, 1692 max running, 304 max pending, priority 0
    3 - /amaltaro_MonteCarlo_eff_ScaleTest_TEST_160829_100904_2175/Production
    997 - /amaltaro_MonteCarlo_HighPrio_ScaleTest_TEST_160829_133434_146/Production
```

meaning, resources are completely filled by the highprio jobs.
Having a look at condor:
```
cmst1@vocms0126:/data/srv/wmagent/current $ condor_q -constraint 'MATCH_EXP_JOBGLIDEIN_CMSSite == "'T1_ES_PIC'"' -af:h WMAgent_RequestName | uniq -c
    701 amaltaro_MonteCarlo_HighPrio_ScaleTest_TEST_160829_133434_146
cmst1@vocms0126:/data/srv/wmagent/current $ 
cmst1@vocms0126:/data/srv/wmagent/current $ condor_q -constraint 'MATCH_EXP_JOBGLIDEIN_CMSSite == "'T2_US_MIT'"' -af:h WMAgent_RequestName | uniq -c
    589 amaltaro_MonteCarlo_HighPrio_ScaleTest_TEST_160829_133434_146
```

only highprio jobs running as well.

What I want to say is, yes, there will be a delay to ramp up the higher priority jobs in the pool, but this is resumed to a few hours now.

**Comment by ticoann** [comment]
It might be handling edge cases. 
The case this addresses is that 

1. If there are low priority workflows with a lot of site white list and high priority has limited site white list. since we are selecting pending job location randomly (actually first in order), high priority job can be blocked submission for a long time. 

2. basically this implementation allows the backdoor for high priority jobs. Which LQ has it.

3. This could potentially create the problem of too many pending jobs in condor if requestor is keep adding higher priority jobs in order.

I think it makes sense to have this back door if 1. is legitimate case (happens more than rare occasion)

But lets see first whether this case happens.

**Comment by amaltaro** [comment]
Seangchan, I agree it would give a faster response for high priority jobs. However, it adds quite a bit of complexity and it's not easy to predict and debug these cases.

> 1. If there are low priority workflows with a lot of site white list and high priority has limited site white list. since we are selecting pending job location randomly (actually first in order), high priority job can be blocked submission for a long time. 

This "long time" noticed in my tests this week was around 1h to get 100 jobs in condor, 2 or 3h to be using half'ish of the pending thresholds and around 10h to be fully using the pending site thresholds for the high prio jobs.

I'm in favor of keeping it unhold for a few weeks. IF, we are proven the system is not really managing highprio work fast enough (<12h), then we can make these changes IMO.

**Comment by vlimant** [comment]
I'd like the situation where high priority jobs are considered always first.

I hear the issue of complexity, and if you have confidence we are down to a couple of hours to surface a higher priority workflow with <4 sites in whitelist, with a monster (1B) jobs in total low priority workflow that can run at >30 sites, then we are good. Otherwise, complexity it should be.

**Comment by hufnagel** [comment]
There is complexity that solves your problem. There is a complexity that is just a layer of band aids where every band aid solves one problem and introduces two new problems. I don't mind the former if warranted, I try to avoid the latter.

**Comment by ticoann** [comment]
@vlimant, Hi Jean-Roch, I think a couple of hours for high priority job to get in when all the sites are pull is too hard requirement with current design. Unless all the jobs are finishing <1h. 
How long is too long to wait high priority jobs start to run when new high priority workflows are assigned?

I discussed with Alan today about the issue. I think what Alan suggested is reasonable. 
Even with the example case above, I think it would be very rare to make the high priority workflow get stuck. (The low priority jobs assigned (although not necessary to run on those sites) in 4 sites has to be pending long time while jobs in other sites finishes first and low priority jobs get back in.

Of course if this is a real problem and happens more often than very very rare, we need to implement even though the complexity of priority handling increase. But I think we should identify the problem first as Alan pointed.

Anyway, I will look at the code in JobSubmitter, whether we can make code cleaner with some refactoring if we have to add this feature. But we the patches we added so far, the situation should be improved unless this particular case with worst scenario happens.
